### ISSUE TITLE
Correct Way to Add Kernel Flags?

### ISSUE THREAD

**Issue by dm17** [zbm-dev/zfsbootmenu#174]
This doesn't seem to work:
https://www.reddit.com/r/voidlinux/comments/dze2bx/how_to_set_kernel_module_parameters/

I can see the flags I've set are not being picked up by 'systool -vm amdgpu'.
I did the following:
1) created /etc/dracut.conf.d/kernelcmdline.conf
2) added these two lines: 
hostonly="yes"
kernel_cmdline="amd_iommu=off"
3) add this line to /etc/modprobe.d/amd.conf:
options amdgpu amd_iommu=off
3) xbps-reconfigure -f linux5.11
4) reboot into 5.11 and see the wrong values for flags using systool.

If I enter ZBM, CTRL+E for edit KCL, and append "amd_iommu=off" (or any flag I try) to the end of the line, then it doesn't seem to boot / just gives a black screen.

Thanks!

**Comment by ahesford** [comment]
1. The dracut `kernel_cmdline` option is only used when you use `dracut --uefi` to build a UEFI bundle. If you're using ZFSBootMenu, the preferred way to set kernel command-line arguments is by assigning the command-line arguments you want in the `org.zfsbootmenu:commandline` property on the BE (or an ancestor of the BE; the property will be inherited). https://github.com/zbm-dev/zfsbootmenu/blob/8886db347a0354fabbf7773a044c17a10e246311/pod/zfsbootmenu.7.pod#zfs-properties
2. `modinfo amdgpu` shows no `amd_iommu` option.
3. You probably want to boot ZFSBootMenu itself with `nomodeset` to avoid the ZBM kernel from attempting to initialize the graphics hardware. That should prevent you from having to mess with the GPU in your final BE.

**Comment by dm17** [comment]
@ahesford Thanks for the speedy reply!
So the way to add to ```org.zfsbootmenu:commandline``` is in the ```Kernel: CommandLine:``` section of ```/etc/zfsbootmenu/config.yaml```? I will try adding ```nomodeset``` there - however, adding amd_iommu=off there had no effect (I checked using ```systool -vm amdgpu```).

**Comment by ahesford** [comment]
"BE" is boot environment, the filesystem you intend to boot.

`Kernel: CommandLine` in `/etc/zfsbootmenu/config.yaml` is only used when you set `EFI: Enabled: true` in the ZFSBootMenu configuration, and that only affects the command-line that is used to boot the ZFSBootMenu image as a UEFI bundle. When setting `EFI: Enabled: true`, you should probably also set `EFI: Versions: false` and you 'll get a nice unversioned file like `/boot/efi/EFI/void/vmlinuz.EFI` (this will depend on where you set `EFI: ImageDir:` and the kernel prefix) plus, if you run `generate-zbm` multiple times, a `vmlinuz-backup.EFI` alongside it that contains the previous version. Just set your PC to boot that EFI file and it will contain the `nomodeset` you added to `Kernel: CommandLine`. If you are using the rEFInd directions from the wiki, you want to add `nomodeset` to the rEFInd configuration instead.

You set `org.zfsbootmenu:commandline` using [`zfs(8)`](https://man.voidlinux.org/zfs.8) command. For example,
```
zfs set org.zfsbootmenu:commandline="amd_iommu=off loglevel=4 consoleblank=600" zpool/ROOT/void
```
if you are booting `zpool/ROOT/void` and you want the screen to blank after 10 minutes and have some amount of kernel logging.

**Comment by dm17** [comment]
Thanks; I got that working - I had an assumption about the ```zfs``` command being irrelevant for non-zfs kernel stuff. I'm not sure how I'd make that more obvious to others - or if it was just my own brain fart :)

I was just going to ask you to go into detail about "you probably want to boot ZFSBootMenu itself with nomodeset to avoid the ZBM kernel from attempting to initialize the graphics hardware..." When I tried it and it fixed the ACPI/AMD/sleep-related bug I just spent loads of time trying to solve! Perhaps ```nomodeset``` should become a default for the sake of reliability & decoupling ZFS-related stuff from loading drivers properly? I'll follow up here with the full list of fixes so others who are banging their heads on the table might figure it out more easily.

@ahesford & zdykstra are legends!
